### ISSUE TITLE
cannot run tests

### ISSUE THREAD

**Issue by OmbraDiFenice** [django-guardian/django-guardian#557]
Hi, I'm trying to run the tests for guardian, but I keep getting an error saying that I'm missing the table 'guardian_post', even with a fresh installation and project.

## Environment

- tried with both django 1.10.4 and 2.0.3
- django-guardian 1.4.9 installed form pip
- python 3.6.4

## Steps to reproduce

1. run django-admin startproject a_project && cd a_project
2. add 'guardian' to INSTALLED_APPS
3. run ./manage.py test guardian -v3

### output:

Creating test database for alias 'default' ('file:memorydb_default?mode=memory&cache=shared')...
Operations to perform:
  Synchronize unmigrated apps: messages, staticfiles
  Apply all migrations: admin, auth, contenttypes, guardian, sessions
Running pre-migrate handlers for application admin
Running pre-migrate handlers for application auth
Running pre-migrate handlers for application contenttypes
Running pre-migrate handlers for application sessions
Running pre-migrate handlers for application guardian
Synchronizing apps without migrations:
  Creating tables...
    Running deferred SQL...
Running migrations:
  Applying contenttypes.0001_initial... OK (0.016s)
  Applying auth.0001_initial... OK (0.042s)
  Applying admin.0001_initial... OK (0.039s)
  Applying admin.0002_logentry_remove_auto_add... OK (0.041s)
  Applying contenttypes.0002_remove_content_type_name... OK (0.099s)
  Applying auth.0002_alter_permission_name_max_length... OK (0.039s)
  Applying auth.0003_alter_user_email_max_length... OK (0.039s)
  Applying auth.0004_alter_user_username_opts... OK (0.041s)
  Applying auth.0005_alter_user_last_login_null... OK (0.033s)
  Applying auth.0006_require_contenttypes_0002... OK (0.006s)
  Applying auth.0007_alter_validators_add_error_messages... OK (0.033s)
  Applying auth.0008_alter_user_username_max_length... OK (0.043s)
  Applying auth.0009_alter_user_last_name_max_length... OK (0.035s)
  Applying guardian.0001_initial... OK (0.149s)
  Applying sessions.0001_initial... OK (0.010s)
Running post-migrate handlers for application admin
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Running post-migrate handlers for application auth
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Running post-migrate handlers for application contenttypes
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Running post-migrate handlers for application sessions
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Running post-migrate handlers for application guardian
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Adding permission 'Permission object (None)'
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/sqlite3/base.py", line 303, in execute
    return Database.Cursor.execute(self, query, params)
sqlite3.OperationalError: no such table: guardian_post

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./manage.py", line 15, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 371, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 365, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/commands/test.py", line 26, in run_from_argv
    super().run_from_argv(argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 288, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 335, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/commands/test.py", line 59, in handle
    failures = test_runner.run_tests(test_labels)
  File "/usr/local/lib/python3.6/site-packages/django/test/runner.py", line 601, in run_tests
    old_config = self.setup_databases()
  File "/usr/local/lib/python3.6/site-packages/django/test/runner.py", line 548, in setup_databases
    self.parallel, **kwargs
  File "/usr/local/lib/python3.6/site-packages/django/test/utils.py", line 176, in setup_databases
    serialize=connection.settings_dict.get('TEST', {}).get('SERIALIZE', True),
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/base/creation.py", line 76, in create_test_db
    self.connection._test_serialized_contents = self.serialize_db_to_string()
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/base/creation.py", line 119, in serialize_db_to_string
    serializers.serialize("json", get_objects(), indent=None, stream=out)
  File "/usr/local/lib/python3.6/site-packages/django/core/serializers/__init__.py", line 128, in serialize
    s.serialize(queryset, **options)
  File "/usr/local/lib/python3.6/site-packages/django/core/serializers/base.py", line 80, in serialize
    for count, obj in enumerate(queryset, start=1):
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/base/creation.py", line 116, in get_objects
    yield from queryset.iterator()
  File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 336, in _iterator
    yield from self._iterable_class(self, chunked_fetch=use_chunked_fetch, chunk_size=chunk_size)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/query.py", line 53, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/usr/local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1066, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 68, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 77, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.6/site-packages/django/db/backends/sqlite3/base.py", line 303, in execute
    return Database.Cursor.execute(self, query, params)
django.db.utils.OperationalError: no such table: guardian_post

Any help?

**Comment by ad-m** [comment]
You can not test a library using a different project than the [one that is dedicated to it](https://github.com/django-guardian/django-guardian/tree/devel/example_project).

In our project, we carry out tests of our library. In your project you test your code, integrations of our library in your code, but not our library itself.

**Comment by OmbraDiFenice** [comment]
Fair enough, I thought it had to be something naive.
I misunderstood the [doc](https://django-guardian.readthedocs.io/en/stable/develop/testing.html#running-tests), anyway I ended up here because I'm having trouble run tests in my project. I get a foreign key error and the last thing I can see is that django is creating tables in the test db related to guardian:

...
Synchronizing apps without migrations:
  Creating tables...
    Creating table guardian_userobjectpermission
    Creating table guardian_groupobjectpermission
    Running deferred SQL...
Traceback (most recent call last):
...

I'll try to look better into it

Thank you

**Comment by ad-m** [comment]
Thank you for reporting. It needs updating!

Why this mark guardian as app without migration? [There is migration.](https://github.com/django-guardian/django-guardian/tree/devel/guardian/migrations)

**Comment by OmbraDiFenice** [comment]
I know, and I don't understand why too... in the "production" db the migration is applied, i can see it inside django_migrations table.
My project has 4 databases defined in settings, could it be because of this?

**Comment by ad-m** [comment]
I have never tested this configuration. I suggest you customize the example_project and check.

**Comment by OmbraDiFenice** [comment]
I think it was me messing up with the code. For some reason I had some (probably old) version of guardian as app folder inside my project, like any other app.

Now I got rid of the local guardian app and used it via pip and everything works as expected.

I tried to copy the "guardian" folder from the standard site-packages directory back inside the project folder and I got the "table 'guardian_post' doesn't exist"error again. I guess that having a local guardian app made Django think that "guardian.testapp" (or any other file inside it) was a regular test module, thus trying to load and run it when I ran the _./manage.py test_ command, [probably since the name starts with "test*"](https://docs.python.org/3/library/unittest.html#test-discovery)

It was silly of me to make such a mistake, at least I hope this can help someone else in future :)
